From 1d86a3f2b660c986c66124532e97778e9f6e2818 Mon Sep 17 00:00:00 2001
From: Qin Jiajia <jiajia.qin@intel.com>
Date: Wed, 4 Mar 2015 03:50:49 -0800
Subject: [PATCH 12/14] Use vaAcquireBufferHandle APIs to instead of
 vaLockBuffer APIs

vaLockBuffer APIs are obsolete. This change will use vaAcquireBufferHandle APIs
to instead of vaLockBuffer APIs.

VA-API version: >= 0.36 (libva 1.4.0)
---
 third_party/libva/va/va.h         | 101 ++++++++++++++++++++++++++++++++++----
 third_party/libva/va/va_backend.h |  13 +++++
 2 files changed, 105 insertions(+), 9 deletions(-)

diff --git a/third_party/libva/va/va.h b/third_party/libva/va/va.h
index 9455023..532e27a 100644
--- a/third_party/libva/va/va.h
+++ b/third_party/libva/va/va.h
@@ -1852,26 +1852,109 @@ VAStatus vaDestroyBuffer (
     VABufferID buffer_id
 );
 
-/** VA buffer information */
+/** \brief VA buffer information */
 typedef struct {
-    /** Buffer handle */
+    /** \brief Buffer handle */
     uintptr_t           handle;
-    /** Buffer type (See VABufferType). */
+    /** \brief Buffer type (See \ref VABufferType). */
     uint32_t            type;
     /**
-     * Buffer memory type (See VASurfaceAttribMemoryType).
+     * \brief Buffer memory type (See \ref VASurfaceAttribMemoryType).
      *
-     * On input to vaLockBuffer(), this field can serve as a hint to
-     * specify the set of memory types the caller is interested in. On
-     * successful return from vaLockBuffer(), the field is updated
-     * with the best matching memory type.
+     * On input to vaAcquireBufferHandle(), this field can serve as a hint
+     * to specify the set of memory types the caller is interested in.
+     * On successful return from vaAcquireBufferHandle(), the field is
+     * updated with the best matching memory type.
      */
     uint32_t            mem_type;
-    /** Size of the underlying buffer. */
+    /** \brief Size of the underlying buffer. */
     size_t              mem_size;
 } VABufferInfo;
 
 /**
+ * \brief Acquires buffer handle for external API usage
+ *
+ * Locks the VA buffer object \ref buf_id for external API usage like
+ * EGL or OpenCL (OCL). This function is a synchronization point. This
+ * means that any pending operation is guaranteed to be completed
+ * prior to returning from the function.
+ *
+ * If the referenced VA buffer object is the backing store of a VA
+ * surface, then this function acts as if vaSyncSurface() on the
+ * parent surface was called first.
+ *
+ * The \ref VABufferInfo argument shall be zero'ed on input. On
+ * successful output, the data structure is filled in with all the
+ * necessary buffer level implementation details like handle, type,
+ * memory type and memory size.
+ *
+ * Note: the external API implementation, or the application, can
+ * express the memory types it is interested in by filling in the \ref
+ * mem_type field accordingly. On successful output, the memory type
+ * that fits best the request and that was used is updated in the \ref
+ * VABufferInfo data structure. If none of the supplied memory types
+ * is supported, then a \ref VA_STATUS_ERROR_UNSUPPORTED_MEMORY_TYPE
+ * error is returned.
+ *
+ * The \ref VABufferInfo data is valid until vaReleaseBufferHandle()
+ * is called. Besides, no additional operation is allowed on any of
+ * the buffer parent object until vaReleaseBufferHandle() is called.
+ * e.g. decoding into a VA surface backed with the supplied VA buffer
+ * object \ref buf_id would fail with a \ref VA_STATUS_ERROR_SURFACE_BUSY
+ * error.
+ *
+ * Possible errors:
+ * - \ref VA_STATUS_ERROR_UNIMPLEMENTED: the VA driver implementation
+ *   does not support this interface
+ * - \ref VA_STATUS_ERROR_INVALID_DISPLAY: an invalid display was supplied
+ * - \ref VA_STATUS_ERROR_INVALID_BUFFER: an invalid buffer was supplied
+ * - \ref VA_STATUS_ERROR_UNSUPPORTED_BUFFERTYPE: the implementation
+ *   does not support exporting buffers of the specified type
+ * - \ref VA_STATUS_ERROR_UNSUPPORTED_MEMORY_TYPE: none of the requested
+ *   memory types in \ref VABufferInfo.mem_type was supported
+ *
+ * @param[in] dpy               the VA display
+ * @param[in] buf_id            the VA buffer
+ * @param[in,out] buf_info      the associated VA buffer information
+ * @return VA_STATUS_SUCCESS if successful
+ */
+VAStatus
+vaAcquireBufferHandle(VADisplay dpy, VABufferID buf_id, VABufferInfo *buf_info);
+
+/**
+ * \brief Releases buffer after usage from external API
+ *
+ * Unlocks the VA buffer object \ref buf_id from external API usage like
+ * EGL or OpenCL (OCL). This function is a synchronization point. This
+ * means that any pending operation is guaranteed to be completed
+ * prior to returning from the function.
+ *
+ * The \ref VABufferInfo argument shall point to the original data
+ * structure that was obtained from vaAcquireBufferHandle(), unaltered.
+ * This is necessary so that the VA driver implementation could
+ * deallocate any resources that were needed.
+ *
+ * In any case, returning from this function invalidates any contents
+ * in \ref VABufferInfo. i.e. the underlyng buffer handle is no longer
+ * valid. Therefore, VA driver implementations are free to reset this
+ * data structure to safe defaults.
+ *
+ * Possible errors:
+ * - \ref VA_STATUS_ERROR_UNIMPLEMENTED: the VA driver implementation
+ *   does not support this interface
+ * - \ref VA_STATUS_ERROR_INVALID_DISPLAY: an invalid display was supplied
+ * - \ref VA_STATUS_ERROR_INVALID_BUFFER: an invalid buffer was supplied
+ * - \ref VA_STATUS_ERROR_UNSUPPORTED_BUFFERTYPE: the implementation
+ *   does not support exporting buffers of the specified type
+ *
+ * @param[in] dpy               the VA display
+ * @param[in] buf_id            the VA buffer
+ * @return VA_STATUS_SUCCESS if successful
+ */
+VAStatus
+vaReleaseBufferHandle(VADisplay dpy, VABufferID buf_id);
+
+/**
  * Locks buffer for external API usage.
  *
  * Locks the VA buffer object buf_id for external API usage like
diff --git a/third_party/libva/va/va_backend.h b/third_party/libva/va/va_backend.h
index 150f8ef..844774f 100644
--- a/third_party/libva/va/va_backend.h
+++ b/third_party/libva/va/va_backend.h
@@ -392,6 +392,19 @@ struct VADriverVTable
                 VASurfaceID surface
         );
 
+        VAStatus
+        (*vaAcquireBufferHandle)(
+            VADriverContextP    ctx,
+            VABufferID          buf_id,         /* in */
+            VABufferInfo *      buf_info        /* in/out */
+        );
+
+        VAStatus
+        (*vaReleaseBufferHandle)(
+            VADriverContextP    ctx,
+            VABufferID          buf_id          /* in */
+        );
+
         /* DEPRECATED */
         VAStatus
         (*vaGetSurfaceAttributes)(
-- 
1.9.1

